crates/count-min-sketch/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+//! Count-Min Sketch implementation in Rust
+//! Based on the paper:  http://dimacs.rutgers.edu/~graham/pubs/papers/cmencyc.pdf
+
+use std::{
+    collections::hash_map::DefaultHasher,
+    hash::{Hash, Hasher},
+};
+
+// TODO this is probably not what we want to use as the hash function;
+// should look at the PgAnyElementHashMap for ideas on how to do
+// hashing that cooperates better with Postgres
+#[derive(Clone)]
+struct HashFunction(DefaultHasher);
+
+impl HashFunction {
+    pub fn new() -> Self {
+        Self(DefaultHasher::new())
+    }
+
+    pub fn hash_into_buckets(&self, item: &str, nbuckets: usize) -> usize {
+        let mut hasher = self.0.clone();
+        item.hash(&mut hasher);
+        let hash_val = hasher.finish();
+        (hash_val % (nbuckets as u64)) as usize
+    }
+}
+
+/// The Count-Min Sketch is a compact summary data structure capable
+/// of representing a high-dimensional vector and answering queries on
+/// this vector, in particular point queries and dot product queries,
+/// with strong accuracy guarantees. Such queries are at the core of
+/// many computations, so the structure can be used in order to answer
+/// a variety of other queries, such as frequent items (heavy
+/// hitters), quantile finding, join size estimation, and more. Since
+/// the data structure can easily process updates in the form of
+/// additions or subtractions to dimensions of the vector (which may
+/// correspond to insertions or deletions, or other transactions), it
+/// is capable of working over streams of updates, at high rates.[1]
+///
+/// [1] http://dimacs.rutgers.edu/~graham/pubs/papers/cmencyc.pdf
+pub struct CountMinSketch {
+    counters: Vec<Vec<isize>>,
+    hashfuncs: Vec<HashFunction>,
+    width: usize,
+    depth: usize,
+}
+
+impl CountMinSketch {
+    pub fn new(epsilon: f64, delta: f64) -> Self {
+        assert!(epsilon.lt(&1f64) && epsilon.gt(&0f64));
+        assert!(delta.lt(&1f64) && delta.gt(&0f64));
```

Comment:
Why `delta.lt(&1f64)` rather than `delta < 1.0`? Also consider writing numeric comparisons using only < and ascending left to right so: `0.0 < delta && delta < 1.0`

Reply:
I think this happened because I had the two lines below in view as I wrote these and having to use `.exp()` and `.ln()` there made this the subconscious choice. 

Unquestionably, `0.0 < delta && delta < 1.0` is better.